libecho/src/Echo.cc

```diff
@@ -37,31 +39,92 @@ void Echo::send(const std::vector<uint8_t> &buffer) {
         output->enqueueData(encoded.data() + i, std::min(atOnce, encoded.size() - i));
         auto status = output->getStreamStatus();
         qDebug() << "Status: " << status.first << " " << status.second;
-        output->waitForTick();
+        // TODO: i changed something and this line now freezes the playback after a short time
+        // output->waitForTick();
     }
 
     output->waitForState(QAudio::State::IdleState);
 }
 
+static double dft(const char *buffer, int samples, int sampleSize, double ratio) {
+    double re = 0;
+    double im = 0;
+    double angle = 0;
+    const double d_angle = 2.0 * M_PI * ratio;
+    const int mask = (1 << sampleSize) - 1;
+
+    for (int i = 0; i < samples; i++) {
+        double val = *((int *) buffer) & mask;
+        re += val * cos(angle);
+        im += val * sin(angle);
+        buffer += (sampleSize / 8);
+        angle += d_angle;
+    }
+
+    return sqrt(re * re + im * im);
+}
+
+// TODO: make it not use 100%cpu
+void getbuff(AudioInput &audio, int bytes, char *buffer) {
+    while (bytes > 0) {
+        int nread = audio.readBytes(buffer, bytes);
+        buffer += nread;
+        bytes -= nread;
+    }
+}
+
 std::vector<uint8_t> Echo::receive() {
-    auto info = input->getStreamInfo();
-
-    const size_t atOnce = info.periodSize;
-    char *array = new char[atOnce];
-    int idx = 0;
-    const int time = 2000000;
-
-    while (input->getStreamStatus().second < time) {
-        int read = input->readBytes(array, atOnce - idx);
-        idx = idx + read;
-        if (idx == atOnce) {
-            idx = 0;
-            qDebug() << "Process data here - or better: schedule it so this loop is not blocked";
-        }
-        input->waitForTick();
+    static QAudioFormat inputFormat = AudioFormatFactory::getDefaultInputFormat();
+    static AudioInput audio(inputFormat);
```

Comment:
These are already being done in the constructor, you can use class member input instead of your's audio.

Reply:
It would be really nice to do so, that code currently constructs unnecessary objects, and threads connected with them